blockprod/src/detail/mod.rs

```diff
@@ -47,158 +85,291 @@ impl BlockProduction {
-                    }
+            .call({
+                let chain_config = Arc::clone(&self.chain_config);
+
+                move |this| {
+                    let best_block_index = this
+                        .get_best_block_index()
+                        .expect("Best block index retrieval failed in block production");
+
+                    let get_ancestor = |block_index: &BlockIndex, ancestor_height: BlockHeight| {
+                        this.get_ancestor(
+                            &block_index.clone().into_gen_block_index(),
+                            ancestor_height,
+                        )
+                        .map_err(|_| {
+                            PropertyQueryError::GetAncestorError(
+                                GetAncestorError::InvalidAncestorHeight {
+                                    block_height: block_index.block_height(),
+                                    ancestor_height,
+                                },
+                            )
+                        })
+                    };
+
+                    let consensus_data = consensus::generate_consensus_data(
+                        &chain_config,
+                        &best_block_index,
+                        block_timestamp,
+                        best_block_index.block_height().next_height(),
+                        get_ancestor,
+                    );
+                    consensus_data.map(|cons_data| (cons_data, best_block_index))
                 }
-
-                None
             })
             .await?
-            .ok_or(BlockProductionError::FailedToConstructBlock(
-                BlockCreationError::CurrentTipRetrievalError,
-            ))?;
-
-        let (_tx, dummy_rx) = crossbeam_channel::unbounded();
-
-        let block_maker = BlockMaker::new(
-            self.chain_config.clone(),
-            self.chainstate_handle.clone(),
-            self.mempool_handle.clone(),
-            self.time_getter.clone(),
-            reward_destination,
-            current_tip_id,
-            current_tip_height,
-            dummy_rx,
-        );
+            .map_err(BlockProductionError::FailedConsensusInitialization)?;
 
-        Ok(block_maker.generate_block(transactions).await?)
+        Ok(consensus_data)
     }
-}
 
-#[cfg(test)]
-mod tests {
-    use tokio::time::{timeout, Duration};
+    fn solve_block(
+        chain_config: Arc<ChainConfig>,
+        mut block_header: BlockHeader,
+        current_tip_height: BlockHeight,
+        stop_flag: Arc<AtomicBool>,
+    ) -> Result<BlockHeader, BlockProductionError> {
+        // TODO: use a separate executor for this loop to avoid starving tokio tasks
+        consensus::finalize_consensus_data(
+            &chain_config,
+            &mut block_header,
+            current_tip_height,
+            stop_flag,
+        )?;
 
-    use super::*;
-    use crate::tests::setup_blockprod_test;
+        Ok(block_header)
```

Comment:
As `finalize_consensus_data()` internally doesn't check the return of `mine()`'s success, the spawned thread could send invalid block headers back to the main thread.

Reply:
Fixed